File: lang/python/material/binary_search.py
def search(ok:int,ng:int,f:bool)->int:
    """二分探索を行う関数
    単調増加の範囲においてokのうちいちばんngに近いものの値を返す
    lambda i:a[i]<x xを含まない最大のiを返す
    lambda i:a[i]<=x xを含む最大のiを返す

    Args:
        ok (int): 評価関数fに渡したときに必ずTrueを返すことが保証されている値
        ng (int): 評価関数fに渡したときに必ずFalseを返すことが保証されている値
        f (bool): 評価関数(引数1/戻り値bool)

    Returns:
        int: 結果
    """
    while 1<abs(ok-ng):
        mid=(ng+ok)//2
        if f(mid):
            ok=mid
        else:
            ng=mid
    return ok

File: modules/cpp/binarySearch.cpp
#include <cstdint>
#include <cmath>
/**
 * 二分探索関数
 * 評価関数evalがtrueになる範囲とfalseになる範囲の境界値を求める
 * @param ok trueになる片極端
 * @param ng falseになる片極端
 * @param eval 評価関数
 */
template <class Func>
int64_t search(int64_t ok, int64_t ng, Func eval)
{
    while (1 < std::abs(ok - ng))
    {
        int64_t mid = (ok + ng) / 2;
        if (eval(mid))
        {
            ok = mid;
        }
        else
        {
            ng = mid;
        }
    }
    return ok;
}